### ISSUE TITLE
Support BatchToSpaceND and SpaceToBatchND

### ISSUE THREAD

**Issue by nsthorat** [tensorflow/tfjs#161]
These are used for atrous convolutions, which is important for mobile devices. We should add these ops / support them in the converter.

**Comment by jgartman** [comment]
@nsthorat I was looking for something to do, I'll try `BatchToSpaceND` if no one has taken it up yet.

**Comment by dikatok** [comment]
Is this still open? I am interested in contributing to implement either of them since it will help converting my model as well XD

**Comment by ManrajGrover** [comment]
@dikatok I don't think anyone is working on `SpaceToBatchND`. Feel free to take it up.

**Comment by dikatok** [comment]
@ManrajGrover cool, thanks

**Comment by nsthorat** [comment]
Ah, let me know if you can't find the workaround. We may have to add support for 5D tensors, which actually shouldn't be too bad. There's another person who also wants support for 5D.

I filed an issue about 5D tensors if somebody wants to try this here: https://github.com/tensorflow/tfjs/issues/249

**Comment by jgartman** [comment]
@nsthorat In the typical case the intermediate tensor I mentioned would actually be 6D.  Based on your comments in the 5D tensor support issue I think I'd be able pull off adding 6D tensors.  However, let me know if it'd be better to focus on other options rather than pursue 6D support at this time.  The tensor returned from this op would usually be 4D after the final reshape so adding 6D tensors just to support an intermediate step seems like it would be a pretty limited use case.

**Comment by nsthorat** [comment]
I think if you've thought about it and there's no way to do it without 6D Tensors without bending backwards, then it's fine to add 6D sampler support.

To be clear, we do support arbitrary rank Tensors everywhere but the texture samplers.

**Comment by jgartman** [comment]
Got it, thanks Nikhil

**Comment by jgartman** [comment]
Just an update in case anyone is following along with this issue.  When https://github.com/tensorflow/tfjs-core/pull/1022 gets merged I'll open up .a PR for a 6D sampler.  After that these ops should be easy to support and I'll open a separate PR for `batchToSpaceND`.

**Comment by jgartman** [comment]
@zaidalyafeai Ya, I  think I have it done for the most part.  Since it's so similar to your PR I was just going to wait until yours gets reviewed.  Then I can take any comments from the review for 5D into account before I open another PR.

**Comment by classner** [comment]
+1, would be very helpful to have this!

**Comment by jgartman** [comment]
The PR I linked to in the last update above is still in review so support for these ops is still waiting on that.  It's hard to give an estimate on time because it really just depends on how long it takes to get PR's merged but I don't think should be too much longer.

**Comment by zaidalyafeai** [comment]
@jgartman, what operations do you need for 5d tensors? For the time being I only implemented and tested `tile` since I need it for https://github.com/tensorflow/tfjs/issues/250. Maybe we can support the most important gpu operations for the time being? just to save now and then finalize the rest later.

**Comment by jgartman** [comment]
@zaidalyafeai I'm only going to need `transpose`.

**Comment by zaidalyafeai** [comment]
Good news the PR for tensors rank 5 is merged @jgartman. Let me know if you face any issue.

**Comment by jgartman** [comment]
Awesome, thanks @zaidalyafeai

**Comment by rodrigob** [comment]
Things seems to be moving forward. What is the status on this feature ?

**Comment by rodrigob** [comment]
tensorflow/tfjs-core#1121 seems to aim at providing BatchToSpaceND, is anyone working on the symmetric  SpaceToBatchND ?

**Comment by dikatok** [comment]
@rodrigob I am currently working on it, sorry for the delay, I was busy with personal matters for the past month and have lost a family member recently. I will try to create an initial PR tomorrow.

**Comment by dsmilkov** [comment]
Hi @dikatok ,

My deepest condolences for your family's loss.

Let us know if you are still interested in pursuing this. It's completely ok if you don't have the time - @jgartman can take on the op. We wouldn't normally be pushing for this, but there is an internal team at Google that depends on this op being added.

Thank you!

**Comment by nsthorat** [comment]
@jgartman would you be willing to take this up now?

@dikatok, sending best wishes. If / when you want to contribute, let us know, and we'd be happy to give you something high-impact.

**Comment by jgartman** [comment]
@nsthorat sure, just a heads up though, it would probably take me through the end of the weekend to get it done

**Comment by nsthorat** [comment]
Yep, no worries! :)

**Comment by dikatok** [comment]
tensorflow/tfjs-core@bf22fc3

**Comment by dikatok** [comment]
Sorry for the delay, still missing the docs though

**Comment by dsmilkov** [comment]
Thanks @dikatok . Looks great, left 3 small comments - happy to merge afterwards!

**Comment by rodrigob** [comment]
With tensorflow/tfjs-core#1176 and tensorflow/tfjs-core#1121 merged. Seems that the only thing missing is to expose these in tfjs-converter (and try converting a model with dilated convolutions).

@pyu10055  could you look into that ?

**Comment by dikatok** [comment]
@rodrigob I have free time at the moment, I'll take it if it's still available.

**Comment by pyu10055** [comment]
@rodrigob yes, I am working on that, PR will be out today.

**Comment by rodrigob** [comment]
Awesome ! Thanks @pyu10055 !
(@dikatok  seems that the task is already taken; thanks a lot for your contributions up to now)

**Comment by dikatok** [comment]
@rodrigob it's cool man, no worries.

**Comment by flyfj** [comment]
seem the changes haven't been added to converter. would love to see it to support converting more models.

**Comment by rodrigob** [comment]
According to the history at
https://github.com/tensorflow/tfjs-converter/commits/master

tfjs-converter version  0.5.4 should already support these operations.
Are you sure you updated the converter ?

**Comment by nsthorat** [comment]
This is now fixed, thanks for all the hard work everyone!
